pubsub/README.md

```diff
@@ -149,22 +156,172 @@ economics (see e.g.
 and
 [here](https://ethresear.ch/t/improving-the-ux-of-rent-with-a-sleeping-waking-mechanism/1480)).
 
+## Message Identification
+
+Pubsub requires to uniquely identify messages via a message ID. This enables
+a wide range of processes like de-duplication, tracking, scoring,
+circuit-breaking, and others.
+
+**The `message_id` is calculated from the `Message` struct.**
+
+By default, **origin-stamping** is in force. This strategy relies on the string
+concatenation of the `from` and `seqno` fields, to uniquely identify a message
+based on the *author*.
+
+Alternatively, a user-defined `message_id_fn` may be supplied, where 
+`message_id_fn(Message) => message_id`. Such a function could compute the hash
+of the `data` field within the `Message`, and thus one could reify
+**content-addressed messaging**.
+
+If fabricated collisions are not a concern, or difficult enough within the
+window the message is relevant in, a `message_id` based on a short digest of
+inputs may benefit performance.
+
+> **[[ Margin note ]]:** There's a potential caveat with using hashes instead of
+> seqnos: the peer won't be able to send identical messages (e.g. keepalives)
+> within the timecache interval, as they will get treated as duplicates. This
+> consequence may or may not be relevant to the application at hand.
+> Reference: [#116](https://github.com/libp2p/specs/issues/116).
+
+**Note that the availability of these fields on the `Message` object will depend
+on the [signature policy](#signature-policy) configured for the topic.**
+
+Whichever the choice, it is crucial that **all peers** participating in a topic
+implement identical message ID calculation logic, or the topic may function
+suboptimally.
```

Comment:
Maybe use stronger wording here? It's not just suboptimal, it's nearly a _malfunction_: divergent message id functions may cause the gossipsub router to use excessive bandwidth and penalize peers unduly for not following up on advertisements.

Reply:
Done.